pkg/integrations/elasticsearch_exporter/elasticsearch_exporter.go

```diff
@@ -0,0 +1,160 @@
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultConfig
+
+	type plain Config
+	return unmarshal((*plain)(c))
+}
+
+func (c Config) Name() string {
+	return "elasticsearch_exporter"
+}
+
+func (c Config) CommonConfig() config.Common {
+	return c.Common
+}
+
+// NewIntegration creates the elasticsearch integration by replicating the main() function of github.com/justwatchcom/elasticsearch_exporter
+// but using yaml configuration instead of kingpin flags.
+func (c Config) NewIntegration(logger log.Logger) (integrations.Integration, error) {
+	if c.URI == "" {
+		return nil, fmt.Errorf("empty URI provided")
+	}
+	esURL, err := url.Parse(c.URI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse es.uri: %w", err)
+	}
+
+	// returns nil if not provided and falls back to simple TCP.
+	tlsConfig := createTLSConfig(c.CA, c.ClientCert, c.ClientPrivateKey, c.InsecureSkipVerify)
+
+	httpClient := &http.Client{
+		Timeout: c.Timeout,
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+	}
+
+	clusterInfoRetriever := clusterinfo.New(logger, httpClient, esURL, c.ExportClusterInfoInterval)
+
+	collectors := []prometheus.Collector{
+		clusterInfoRetriever,
+		collector.NewClusterHealth(logger, httpClient, esURL),
+		collector.NewNodes(logger, httpClient, esURL, c.AllNodes, c.Node),
+	}
+
+	if c.ExportIndices || c.ExportShards {
+		iC := collector.NewIndices(logger, httpClient, esURL, c.ExportShards)
+		collectors = append(collectors, iC)
+		if registerErr := clusterInfoRetriever.RegisterConsumer(iC); registerErr != nil {
+			return nil, fmt.Errorf("failed to register indices collector in cluster info: %w", err)
+		}
+	}
+
+	if c.ExportShards {
+		collectors = append(collectors, collector.NewSnapshots(logger, httpClient, esURL))
+	}
+
+	if c.ExportClusterSettings {
+		collectors = append(collectors, collector.NewClusterSettings(logger, httpClient, esURL))
+	}
+
+	if c.ExportIndicesSettings {
+		collectors = append(collectors, collector.NewIndicesSettings(logger, httpClient, esURL))
+	}
+
+	start := func(ctx context.Context) error {
+		// start the cluster info retriever
+		switch runErr := clusterInfoRetriever.Run(ctx); runErr {
+		case nil:
+			_ = level.Info(logger).Log(
+				"msg", "started cluster info retriever",
+				"interval", c.ExportClusterInfoInterval.String(),
+			)
+		case clusterinfo.ErrInitialCallTimeout:
+			_ = level.Info(logger).Log("msg", "initial cluster info call timed out")
+		default:
+			_ = level.Error(logger).Log("msg", "failed to run cluster info retriever", "err", err)
+			return err
+		}
+
+		// Wait until we're done
+		<-ctx.Done()
+		return ctx.Err()
+	}
+
+	return integrations.NewCollectorIntegration(c.Name(),
+		integrations.WithCollectors(collectors...),
+		integrations.WithRunner(start),
+	), nil
+}
+
+func init() {
+	integrations.RegisterIntegration(&Config{})
+}
```

Comment:
Can you add this to below the last method of the `Config` struct to be consistent with the other files? I was confused when I didn't see this closer to the top :)

Reply:
Moved, and also used pointer receivers and extracted the `New()` constructor to make it consistent in https://github.com/grafana/agent/pull/347/commits/f5a1df4b053cbcae27e1ce83b4d408fe8ef3541d